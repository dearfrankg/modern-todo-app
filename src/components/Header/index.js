import React from 'react';
import { string, func } from 'prop-types';

const propTypes = {
  value: string.isRequired,
  handleSubmit: func.isRequired,
  handleChange: func.isRequired,
};

const Header = ({ value, handleSubmit, handleChange }) => (
  <header className="header">
    <h1>
todos
    </h1>

    <input
      autoFocus
      placeholder="What needs to be done?"
      type="text"
      className="new-todo"
      value={value}
      onChange={handleChange}
      onKeyDown={handleSubmit}
    />
  </header>
);

Header.propTypes = propTypes;

export default Header;
