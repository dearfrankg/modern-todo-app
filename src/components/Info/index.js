import React from "react";

const Info = () => (
  <footer className="info">
    <p>Double-click to edit a todo</p>
    <p>
      Created using by <button href="#">Frank Gutierrez</button>
    </p>
    <p>
      Inspired by <button href="#">SÃ©rgio A. Kopplin</button>
    </p>
  </footer>
);

export default Info;
