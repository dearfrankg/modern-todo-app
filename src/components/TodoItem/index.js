import React from "react";
import { objectOf, any, func, bool, string } from "prop-types";
import classNames from "classnames";

const propTypes = {
  inputRef: func.isRequired,
  todo: objectOf(any).isRequired,
  toggleTodo: func.isRequired,
  editing: bool.isRequired,
  editText: string.isRequired,
  deleteTodo: func.isRequired,
  handleEdit: func.isRequired,
  handleKeyDown: func.isRequired,
  handleChange: func.isRequired,
  handleSubmit: func.isRequired
};

const TodoItem = ({
  inputRef,
  todo,
  toggleTodo,
  editing,
  editText,
  deleteTodo,
  handleEdit,
  handleKeyDown,
  handleChange,
  handleSubmit
}) => (
  <li
    key={todo.id}
    className={classNames({
      completed: todo.completed,
      editing
    })}
  >
    <div className="view">
      <input
        className="toggle"
        type="checkbox"
        checked={todo.completed}
        onChange={() => toggleTodo(todo)}
      />

      <label onDoubleClick={handleEdit}>{todo.text}</label>

      <button className="destroy" onClick={() => deleteTodo(todo.id)} />
    </div>

    <input
      ref={inputRef}
      className="edit"
      value={editText}
      onBlur={handleSubmit}
      onChange={handleChange}
      onKeyDown={handleKeyDown}
    />
  </li>
);

TodoItem.propTypes = propTypes;

export default TodoItem;
