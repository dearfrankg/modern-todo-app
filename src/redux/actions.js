import * as types from 'redux/types';

export const addTodo = text => ({ type: types.ADD_TODO, text });
export const deleteTodo = id => ({ type: types.DELETE_TODO, id });
export const editTodo = (id, text) => ({ type: types.EDIT_TODO, id, text });
export const toggleTodo = todo => ({ type: types.TOGGLE_TODO, todo });
export const completeAll = () => ({ type: types.COMPLETE_ALL });
export const clearCompleted = () => ({ type: types.CLEAR_COMPLETED });
