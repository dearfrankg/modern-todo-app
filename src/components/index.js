import Footer from './Footer';
import Header from './Header';
import Info from './Info';
import TodoItem from './TodoItem';

export {
  Footer, Header, Info, TodoItem,
};
