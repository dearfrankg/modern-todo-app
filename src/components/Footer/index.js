import React from 'react';
import {
  arrayOf, any, func, string,
} from 'prop-types';
import { Link } from 'react-router-dom';

import { getItemsLeftLabel, getListStatus } from 'helpers';
import * as filters from 'redux/types';

const propTypes = {
  todos: arrayOf(any).isRequired,
  clearCompleted: func.isRequired,
  activeFilter: string.isRequired,
};

const Footer = ({ todos, clearCompleted, activeFilter }) => {
  const hasTodos = todos.length > 0;
  const hasCompletedItems = getListStatus(todos).completedItems > 0;

  const footerComponent = (
    <div>
      <span className="todo-count">
        {getItemsLeftLabel(todos)}
      </span>

      <ul className="filters">
        <li>
          <Link className={activeFilter !== filters.SHOW_ALL ? '' : 'selected'} to="#/">
            All
          </Link>
        </li>
        <li>
          <Link className={activeFilter !== filters.SHOW_ACTIVE ? '' : 'selected'} to="#/active">
            Active
          </Link>
        </li>
        <li>
          <Link
            className={activeFilter !== filters.SHOW_COMPLETED ? '' : 'selected'}
            to="#/completed"
          >
            Completed
          </Link>
        </li>
      </ul>

      {hasCompletedItems && (
        <button className="clear-completed" onClick={clearCompleted}>
          Clear completed
        </button>
      )}
    </div>
  );

  return (
    <footer className="footer">
      {hasTodos && footerComponent}
    </footer>
  );
};

Footer.propTypes = propTypes;

export default Footer;
