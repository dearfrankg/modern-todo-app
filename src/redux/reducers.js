import * as types from 'redux/types';

export const initialState = {
  todos: [],
};

const todos = (state = initialState.todos, action) => {
  if (!action) return false;

  switch (action.type) {
    case types.ADD_TODO: {
      const maxId =
        state.length === 0
          ? 0
          : state.reduce((prev, current) => (prev.id > current.id ? prev : current)).id;

      return [
        {
          id: maxId + 1,
          text: action.text,
          completed: false,
        },
        ...state,
      ];
    }

    case types.DELETE_TODO: {
      return state.filter(todo => todo.id !== action.id);
    }

    case types.EDIT_TODO: {
      return state.map((todo) => {
        if (todo.id === action.id) return { ...todo, text: action.text };

        return todo;
      });
    }

    case types.TOGGLE_TODO: {
      return state.map((todo) => {
        if (todo.id === action.todo.id) return { ...todo, completed: !action.todo.completed };

        return todo;
      });
    }

    case types.COMPLETE_ALL: {
      const areAllMarked = state.every(todo => todo.completed);
      return state.map(todo => ({
        ...todo,
        completed: !areAllMarked,
      }));
    }

    case types.CLEAR_COMPLETED: {
      return state.filter(todo => todo.completed === false);
    }

    default:
      return state;
  }
};

export default todos;
