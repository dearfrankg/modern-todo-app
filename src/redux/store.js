import { createStore, combineReducers } from 'redux';
import { routerReducer } from 'react-router-redux';

import rootReducer from 'redux/reducers';
import { loadState, saveState } from 'redux/localStorage';

const persistedState = loadState();

const store = createStore(
  combineReducers({
    todos: rootReducer,
    router: routerReducer,
  }),
  persistedState,
);

store.subscribe(() => {
  saveState({
    todos: store.getState().todos,
  });
});

export default store;
