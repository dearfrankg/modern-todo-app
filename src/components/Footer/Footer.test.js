import React from 'react';
import { shallow } from 'enzyme';

import { Footer } from 'components';
import * as filters from 'redux/types';

const mock = {
  todos: [],
  clearCompleted: () => {},
  activeFilter: '',
};

describe('Footer:', () => {
  it('should render correctly with default props', () => {
    expect(shallow(<Footer {...mock} />)).toMatchSnapshot();
  });

  it('should render correctly SHOW_ALL filter', () => {
    expect(
      shallow(
        <Footer
          {...{
            ...mock,
            activeFilter: filters.SHOW_ALL,
          }}
        />,
      ),
    ).toMatchSnapshot();
  });

  it('should render correctly SHOW_ACTIVE filter', () => {
    expect(
      shallow(
        <Footer
          {...{
            ...mock,
            activeFilter: filters.SHOW_ACTIVE,
          }}
        />,
      ),
    ).toMatchSnapshot();
  });

  it('should render correctly SHOW_COMPLETED filter', () => {
    expect(
      shallow(
        <Footer
          {...{
            ...mock,
            activeFilter: filters.SHOW_COMPLETED,
          }}
        />,
      ),
    ).toMatchSnapshot();
  });

  it('should render correctly when completed items === 0', () => {
    expect(
      shallow(
        <Footer
          {...{
            ...mock,
            todos: [
              {
                id: 1,
                completed: true,
                text: 'text',
              },
            ],
          }}
        />,
      ),
    ).toMatchSnapshot();
  });
});
