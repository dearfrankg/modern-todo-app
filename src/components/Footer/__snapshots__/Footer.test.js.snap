// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer: should render correctly SHOW_ACTIVE filter 1`] = `
<footer
  className="footer"
/>
`;

exports[`Footer: should render correctly SHOW_ALL filter 1`] = `
<footer
  className="footer"
/>
`;

exports[`Footer: should render correctly SHOW_COMPLETED filter 1`] = `
<footer
  className="footer"
/>
`;

exports[`Footer: should render correctly when completed items === 0 1`] = `
<footer
  className="footer"
>
  <div>
    <span
      className="todo-count"
    >
      <span>
        0 items left
      </span>
    </span>
    <ul
      className="filters"
    >
      <li>
        <Link
          className=""
          replace={false}
          to="#/"
        >
          All
        </Link>
      </li>
      <li>
        <Link
          className=""
          replace={false}
          to="#/active"
        >
          Active
        </Link>
      </li>
      <li>
        <Link
          className=""
          replace={false}
          to="#/completed"
        >
          Completed
        </Link>
      </li>
    </ul>
    <button
      className="clear-completed"
      onClick={[Function]}
    >
      Clear completed
    </button>
  </div>
</footer>
`;

exports[`Footer: should render correctly with default props 1`] = `
<footer
  className="footer"
/>
`;
